### PR TITLE
add support for annotation voting

### LINKED ISSUE #110: Support voting on annotations
As proposed by Andrew: https://gist.github.com/ajturner/cd7b927b387b360e4c1f20a0f6a34014

We propose to add the following functions to hub.js:
* add positive or negative vote to an annotation (e.g. `addAnnotationVote`)
* get positive votes for an annotation (e.g. `searchAnnotationUpVotes`)
* get negative votes for an annotation (e.g. `searchAnnotationDownVotes`)

### PR DESCRIPTION
**Pull request by jgravois**
resolves #110 
```js
const orgId = "uCX...";
const authentication = new arcgisRest.UserSession({ username, password })

const serviceUrl = await arcgisHub.getAnnotationServiceUrl(orgId)
const url = `${serviceUrl}/0`;

const searchResponse = await arcgisHub.searchAnnotations({
  url,
  where: "OBJECTID = 1"
});

arcgisHub.voteOnAnnotation({
  url,
  authentication,
  annotation: searchResponse.data[0],
  downVote: true
});
```
to do:
-  [x] tests e5fddb3

### CODE REVIEW

File: packages/annotations/src/vote.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    });
+  }
+
+  return searchAnnotations({
+    url,
+    where: `parent_id = ${annotation.attributes.OBJECTID} AND author = '${
+      auth.username
+    }'`
+  }).then(searchResponse => {
+    // dont let folks vote on an idea more than once
```
Comment (tomwayson) [inline packages/annotations/src/vote.ts]: What if they want to change their vote from an upvote to a downvote? We don't (yet) provide an API to remove the previous vote first, so it seems that unless you know the implementation details you can't really change your vote. Is that intentional?

File: packages/annotations/src/vote.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        const deleteOptions: IDeleteFeaturesRequestOptions = {
+          url,
+          authentication: requestOptions.authentication,
+          objectIds: [queryResponse.features[0].attributes.OBJECTID]
+        };
+        return (
+          deleteFeatures(deleteOptions)
+            // i have NO idea why the union return type is being ignored
+            .then(() => {
+              return { addResults: [{ success: true }] } as IAddFeaturesResult;
```
Comment (jgravois) [inline packages/annotations/src/vote.ts]: in cc576c880617b57cb64443a0fa30d48f4ff0809f i added support for vote switching. this is accomplished by deleting the previous vote (if the new vote is different than the first).

for reasons i don't understand, i had to shape the response of the deleteFeatures() call to impersonate `IAddFeaturesResult` in order to get TypeScript to shut up.

File: packages/annotations/src/layer-definition.ts [hunk 3]
```diff
@@ -202,6 +202,15 @@ export const annotationServiceDefinition: ILayerDefinition = {
...
       defaultValue: null
+    },
+    {
+      name: "value",
+      type: "esriFieldTypeInteger",
+      alias: "value",
+      nullable: true,
+      editable: true,
+      domain: null,
+      defaultValue: null
```
Comment (jgravois) [inline packages/annotations/src/layer-definition.ts]: just to be explicit, this creates a new `value` field in hosted feature services that have yet to be created. i'm not sure what we can/should do about the ones that are already 'in the wild'.

File: packages/annotations/src/vote.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  if (annotation.attributes.author === auth.username) {
+    return new Promise((resolve, reject) => {
+      reject(
+        new ArcGISRequestError(`Users may not vote on their own comment/idea.`)
+      );
+    });
+  }
+
+  // searchAnnotations() would make more xhrs for user metadata
+  return queryFeatures({
```
Comment (jgravois) [inline packages/annotations/src/vote.ts]: i realized during the most recent refactor that `searchAnnotations()` creates even more network chatter (via appending user metadata), so i switched to using `queryFeatures()` instead.

File: packages/annotations/src/vote.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    url,
+    where: `parent_id = ${annotation.attributes.OBJECTID} AND author = '${
+      auth.username
+    }'`
+  }).then((
+    queryResponse: any /* why can't i declare it as IQueryFeaturesResponse ? */
```
Comment (jgravois) [inline packages/annotations/src/vote.ts]: here's another one headscratcher.

`queryFeatures()` returns a union type, yet i can't explicitly cast a response i know will be `IQueryFeaturesResponse` (and not `IQueryResponse`) to introspect properties that i **know** will be available, so i've resorted to yet another `any`.

my TS foo is too weak.

File: packages/annotations/src/layer-definition.ts (hunk 3 above)
Comment (tomwayson) [reply packages/annotations/src/layer-definition.ts]: we should probably create a migrate function

File: packages/annotations/src/vote.ts (hunk 5 above)
Comment (tomwayson) [reply packages/annotations/src/vote.ts]: by cast, did you mean `(queryResponse as IQueryFeaturesResponse)`?

File: packages/annotations/src/layer-definition.ts (hunk 3 above)
Comment (tomwayson) [reply packages/annotations/src/layer-definition.ts]: to be clear, I don't necessarily mean that has to happen in this PR.

File: packages/annotations/src/vote.ts (hunk 5 above)
Comment (jgravois) [reply packages/annotations/src/vote.ts]: yup.  	

> Argument of type 'IQueryFeaturesResponse' is not assignable to parameter of type '(value: IQueryFeaturesResponse | IQueryResponse) => IQueryFeaturesResponse | IQueryResponse | PromiseLike<IQueryFeaturesResponse | IQueryResponse>'.
  Type 'IQueryFeaturesResponse' provides no match for the signature '(value: IQueryFeaturesResponse | IQueryResponse): IQueryFeaturesResponse | IQueryResponse | PromiseLike<IQueryFeaturesResponse | IQueryResponse>

🤔

File: packages/annotations/src/layer-definition.ts (hunk 3 above)
Comment (jgravois) [reply packages/annotations/src/layer-definition.ts]: that's what i assumed.

File: packages/annotations/src/vote.ts (hunk 2 above)
Comment (tomwayson) [reply packages/annotations/src/vote.ts]: it seems to me that after deleting we still need to add the new vote, but you `return` here, so I don't think that happens. Am I missing something?

**Pull request merged**